Fix maxSpinThreads enforcement in spinlock acquire

babsingh opened the pull request:
Use a compare-and-swap loop when incrementing `spinThreads` so that the
`maxSpinThreads` limit is checked and updated atomically.

The previous check-then-increment sequence allowed multiple threads to
observe a value below the limit and increment it concurrently, causing
`spinThreads` to exceed `maxSpinThreads`.

Related: #7859

babsingh commented:
@keithc-ca Could you please review these changes?

fyi @JamesKingdon @jdmpapin

babsingh commented:
jenkins build all

babsingh commented:
jenkins build all

JamesKingdon commented:
Hi, thanks for looking at this.
Doesn't this still leave a race window between the last invocation of the test in omrthread_spinlock_acquire:
`if (oldState == VM_AtomicSupport::lockCompareExchange(target, oldState, newState, true)) {`
and the point where spinThreads is decremented at the end of the method? I was wondering if we would need to add one more invocation of the lockCompareExchange test _after_ decrementing spinThreads to eliminate this.

keithc-ca commented on thread/common/omrthread.c:
```diff
@@ -4360,7 +4360,8 @@ monitor_exit(omrthread_t self, omrthread_monitor_t monitor)
 		omrthread_spinlock_swapState(monitor, J9THREAD_MONITOR_SPINLOCK_UNOWNED);
  		MONITOR_LOCK(monitor, CALLER_MONITOR_EXIT1);
- 		if (0 == monitor->spinThreads) {
+		issueReadWriteBarrier();
+		if (0 == monitor->spinThreads) {
```
I don't expect the barrier to help. Any threads that have determined that they should spin but have not yet incremented `spinThreads` will continue to have the problem this aims to correct.

babsingh commented:
Other than `monitor_enter_three_tier`, the only other caller of `omrthread_spinlock_acquire` is `omrthread_monitor_try_enter_using_threadId`. That path doesn’t have the same behavior; it only attempts to acquire the monitor once, and the caller is expected to retry as needed.

babsingh commented:
@JamesKingdon Could you confirm whether this PR has any impact on the reported issue in https://github.com/eclipse-omr/omr/issues/7859?

> It appears that the test is an optimization to avoid calling `unblock_spinlock_threads` on the basis that a spinning thread will acquire the monitor without the cost of waking a blocked thread. Unfortunately this appears to introduce a race condition between the last iteration in `omrthread_spinlock_acquire` and the point where `spinThreads` is decremented.

As noted in my earlier comments, this race condition may not actually exist, since `omrthread_spinlock_swapState` is invoked after `omrthread_spinlock_acquire`. This suggests the root cause of the issue may lie elsewhere. It is also possible that the reported symptoms result from incorrect monitor usage.

JamesKingdon commented:
Potentially we could, but I'd have to build a new testfix with the change and ask the customer to test with it, and I've already put the customer through multiple rounds of testing debug builds.
We know for sure that bypassing the condition described in #7878 (this change https://github.com/JamesKingdon/openj9-omr/commit/a15cf2f6458b59ba36dce92a475ab6f5d179c4ba#diff-4a9ee7d0885c2a33f954f2b85c301b20fd601a1f97eab4e513200172cfd9e443R4454-R4458)  prevents the hang for the customer - are there any other theories as to how that would be the case?

babsingh commented:
> are there any other theories as to how that would be the case?

Here are a few possible scenarios that could produce similar symptoms:
- Releasing the monitor without holding it
- Entering the monitor twice without releasing
- Double monitor exit without matching enter
- Forgetting to signal after waiting

Have all critical sections that interact with the monitor been reviewed to ensure correct usage and proper enter/exit pairing?

JamesKingdon commented:
Yes, the code has been looked at pretty carefully by the dev team.

babsingh commented:
This fix was developed based on the reported symptoms; however, we haven’t received confirmation that it resolves the issue. The PR cannot be merged until it is verified that the fix fully addresses the problem.

dspath commented:
Our client can recreate the issue.  They are running with a test fix which I'm not sure is a "final" fix or not but they will test a proposed "official" fix if that would be helpful.

keithc-ca commented:
> cannot be merged until the fix is verified to fully address the problem

We should also be sure it doesn't break anything, including having a negative impact on performance.

babsingh commented:
jenkins build all

keithc-ca commented on thread/common/omrthread.c:
```diff
@@ -4365,7 +4365,7 @@ monitor_exit(omrthread_t self, omrthread_monitor_t monitor)
 		omrthread_spinlock_swapState(monitor, J9THREAD_MONITOR_SPINLOCK_UNOWNED);
  		MONITOR_LOCK(monitor, CALLER_MONITOR_EXIT1);
- 		if (0 == monitor->spinThreads) {
+		if (0 == compareAndSwapUDATA((uintptr_t *)&monitor->spinThreads, 0, 0)) {
```
This just seems like a more expensive variant of the barrier in a previous version. The field is volatile, so it's unclear how this does anything new other than hinder performance.

babsingh replied:
Agreed, I've removed the CAS based reads from the unblock paths.

keithc-ca commented on thread/common/threadhelpers.cpp:
```diff
@@ -74,14 +74,18 @@ omrthread_spinlock_acquire(omrthread_t self, omrthread_monitor_t monitor)
 #if defined(OMR_THR_SPIN_WAKE_CONTROL)
 	BOOLEAN spinning = TRUE;
 	if (OMRTHREAD_IGNORE_SPIN_THREAD_BOUND != lib->maxSpinThreads) {
-		if (monitor->spinThreads < lib->maxSpinThreads) {
-			VM_AtomicSupport::add(&monitor->spinThreads, 1);
-		} else {
```
This seems to be the essential improvement. Previously, any number of threads could see `spinThreads < maxSpinThreads` and each increment `spinThreads` with none of them noticing that the line was crossed. However, I think this is missing a check after the atomic increment succeeds that the maximum has not been reached.

babsingh replied:
Do you mean checking for `spinThreads > maxSpinThreads` after the CAS? A successful CAS can produce at most `maxSpinThreads`, since it is attempted only when the observed value is below the limit. Reaching exactly the maximum should still allow that thread to spin.

keithc-ca replied:
I think you're right.

However, I think the change here should be repeated for lines 261-268.

babsingh replied:
The latest push repeats the changes for lines 261-268.

babsingh committed:
Fix maxSpinThreads enforcement in spinlock acquire

Use a compare-and-swap loop when incrementing spinThreads so that the
maxSpinThreads limit is checked and updated atomically.

The previous check-then-increment sequence allowed multiple threads to
observe a value below the limit and increment it concurrently, causing
spinThreads to exceed maxSpinThreads.

Related: #7859

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

keithc-ca reviewed (approved):
I believe this does accomplish the reduced goal of preventing "spinThreads from exceeding maxSpinThreads", and that seems appropriate. Like others, however, I don't see how this has any bearing on the problem of apparent lost notifications.

0xdaryl reviewed (approved):
Approving, relying heavily on @keithc-ca's more thorough review.
